### DISCUSSION TITLE
Restart a specific plugin

### DISCUSSION THREAD

**Discussion by lqd1434** [eclipse-theia/theia#14576] [Q&A]
Is there a way to restart a specific plugin in theia ？
like i can undeploy it and then deploy it

**Comment by msujew** [comment]
Hey @lqd1434,

no I don't think there's a way to do that right now. VS Code allows to restart the whole extension host, because it's architecturally pretty much impossible to restart specific plugins, as they all run in the same process. We don't even have the ability to restart the extension host right now.

What is your use case for that? Maybe there's a different way of accomplishing what you're trying to do.

**Reply by lqd1434** [reply]
Maybe my thinking is wrong. This problem originally came from the fact that when I was using the clangd plugin, I set the clangd configuration through the Preferences schema
`--query-driver=${env:CLANGD_DRIVER}`, and then set this env through envVariableSrver in other places in the frontend, but I found that clangd did not get this value. I guessed that I set this env too late, but later I found that clang provided a command to restart the language server, and after restarting it still could not get this env, so I want to know how these environment variables are resolved
```typescript
export const ClangdSchema = [
    {
        name: 'clangd.path',
        value: '${env:CLANGD}'
    },
    {
        name: 'clangd.arguments',
        value: [
            '--query-driver=${env:CLANGD_DRIVER}',
            '--compile-commands-dir=${workspaceFolder}/.config'
        ]
    }
];
```
In addition, I noticed `env:CLANG`  can resolve correctly

**Reply by msujew** [reply]
I assume that the clangd extension gets this environment variable from the plugin host, whereas you set the environment variable only for the Theia server process. Once the server starts the plugin host process, its environment variables can no longer be changed by changing the `env` of the Theia server process.

You would need some way to explicitly communicate this changed environment variable to the plugin host process. Maybe via a custom extension? By default, only restarting the frontend would result in the backend restarting the plugin host, which would then restart with the new environment variables.

**Reply by lqd1434** [reply]
That sounds right, the env:CLANGD I set in the backend can be resolved correctly.
Thanks for your answer, I will see if there is any other way tomorrow.

**Reply by YathindraNikshipth** [reply]
Hey @msujew, Could you please point the file where the backend starts the plugin host process? Just trying to figure out the flow.

**Reply by msujew** [reply]
@YathindraNikshipth this is the method that spawns the plugin host:
https://github.com/eclipse-theia/theia/blob/16a51862ce91d4c3a68e3f5b1a087e2f7152b3bf/packages/plugin-ext/src/hosted/node/hosted-plugin-process.ts#L172
